src/app/ReadClient.cpp

```diff
@@ -0,0 +1,249 @@
+            // fill to construct event paths
+        }
+        request.EndOfReadRequest();
+        SuccessOrExit(request.GetError());
+
+        err = writer.Finalize(&msgBuf);
+        SuccessOrExit(err);
+
+        VerifyOrExit(msgBuf->EnsureReservedSize(System::PacketBuffer::kDefaultHeaderReserve), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+    }
+
+    ClearExistingExchangeContext();
+
+    mpExchangeCtx = mpExchangeMgr->NewContext({ aNodeId, 0, aAdminId }, this);
+    VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_NO_MEMORY);
+    mpExchangeCtx->SetResponseTimeout(kImMesssageTimeout);
+
+    err = mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReadRequest, std::move(msgBuf),
+                                     Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse));
+    SuccessOrExit(err);
+    MoveToState(ClientState::ReadRequestSending);
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+void ReadClient::OnMessageReceived(Messaging::ExchangeContext * apExchangeContext, const PacketHeader & aPacketHeader,
+                                   const PayloadHeader & aPayloadHeader, System::PacketBufferHandle aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    VerifyOrDie(apExchangeContext == mpExchangeCtx);
+
+    ClearExistingExchangeContext();
+
+    err = ProcessReportData(std::move(aPayload));
+    SuccessOrExit(err);
+
+exit:
+    Reset();
+}
+
+CHIP_ERROR ReadClient::ClearExistingExchangeContext()
+{
+    if (mpExchangeCtx != nullptr)
+    {
+        mpExchangeCtx->Abort();
+        mpExchangeCtx = nullptr;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ReadClient::ProcessReportData(System::PacketBufferHandle aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    ReportData::Parser report;
+
+    bool isEventListPresent  = false;
+    bool suppressResponse    = false;
+    bool moreChunkedMessages = false;
+
+    chip::System::PacketBufferTLVReader reader;
+
+    reader.Init(std::move(aPayload));
+    reader.Next();
+
+    err = report.Init(reader);
+    SuccessOrExit(err);
+
+    err = report.CheckSchemaValidity();
+    SuccessOrExit(err);
+
+    err = report.GetSuppressResponse(&suppressResponse);
+    if (CHIP_END_OF_TLV == err)
+        err = CHIP_NO_ERROR;
+    SuccessOrExit(err);
+
+    err = report.GetMoreChunkedMessages(&moreChunkedMessages);
+    if (CHIP_END_OF_TLV == err)
+        err = CHIP_NO_ERROR;
+    SuccessOrExit(err);
+
+    {
+        EventList::Parser eventList;
+
+        err = report.GetEventDataList(&eventList);
+        if (CHIP_NO_ERROR == err)
+        {
+            isEventListPresent = true;
+        }
+        else if (CHIP_END_OF_TLV == err)
+        {
+            isEventListPresent = false;
+            err                = CHIP_NO_ERROR;
+        }
+        SuccessOrExit(err);
+
+        if (isEventListPresent && nullptr != mpDelegate && !moreChunkedMessages)
+        {
+            // re-initialize the reader (reuse to save stack depth).
```

Comment:
So this assumes nothing will need to use the reader after this point, right?

This seems really fragile.  I'd prefer we use a new reader and file a followup issue to re-evaluate once this code is more fleshed-out and stable whether reuse makes sense then.

Reply:
done, thanks